src/core/actions/cityactions/Build.java

```diff
@@ -1,46 +1,214 @@
+                        if(board.getTerrainAt(targetPos.x, targetPos.y) != goodTerrain){
+                            return false;
+                        }
+                    }
+                    return true;
+                }
+            case LUMBER_HUT:
+                if(stars >= TribesConfig.LUMBER_HUT_COST && t.isResearched(Types.BUILDING.LUMBER_HUT.getTechnologyRequirement())){
+                    for(Types.TERRAIN goodTerrain : Types.BUILDING.LUMBER_HUT.getTerrainRequirements()){
+                        if(board.getTerrainAt(targetPos.x, targetPos.y) != goodTerrain){
+                            return false;
+                        }
+                    }
+                    return true;
+                }
+            case TEMPLE:
+                if(stars >= TribesConfig.TEMPLE_COST && t.isResearched(Types.BUILDING.TEMPLE.getTechnologyRequirement())){
+                    for(Types.TERRAIN goodTerrain : Types.BUILDING.TEMPLE.getTerrainRequirements()){
+                        if(board.getTerrainAt(targetPos.x, targetPos.y) != goodTerrain){
+                            return false;
+                        }
+                    }
+                    return true;
+                }
+            case WATER_TEMPLE:
+                if(stars >= TribesConfig.TEMPLE_COST && t.isResearched(Types.BUILDING.WATER_TEMPLE.getTechnologyRequirement())){
+                    for(Types.TERRAIN goodTerrain : Types.BUILDING.WATER_TEMPLE.getTerrainRequirements()){
+                        if(board.getTerrainAt(targetPos.x, targetPos.y) != goodTerrain){
+                            return false;
+                        }
+                    }
+                    return true;
+                }
+            case MOUNTAIN_TEMPLE:
+                if(stars >= TribesConfig.TEMPLE_COST && t.isResearched(Types.BUILDING.MOUNTAIN_TEMPLE.getTechnologyRequirement())){
+                    for(Types.TERRAIN goodTerrain : Types.BUILDING.MOUNTAIN_TEMPLE.getTerrainRequirements()){
+                        if(board.getTerrainAt(targetPos.x, targetPos.y) != goodTerrain){
+                            return false;
+                        }
+                    }
+                    return true;
+                }
+            case FOREST_TEMPLE:
+                if(stars >= TribesConfig.TEMPLE_FOREST_COST && t.isResearched(Types.BUILDING.FOREST_TEMPLE.getTechnologyRequirement())){
+                    for(Types.TERRAIN goodTerrain : Types.BUILDING.FOREST_TEMPLE.getTerrainRequirements()){
+                        if(board.getTerrainAt(targetPos.x, targetPos.y) != goodTerrain){
+                            return false;
+                        }
+                    }
+                    return true;
+                }
+        }
         return false;
     }
 
     @Override
     public boolean execute(GameState gs) {
-        //TODO: Executes the action
+        //TODO: Add monuments, roads
```

Comment:
There are some side effects for buildings that are not counted here (computed in City.java - addBuildings, setProduction and changeProduction). Those methods were written by Judy, I'm not sure which ones should be called. Maybe just leave a TODO for the moment there.